datasync/chaindatafetcher/kas/repository_traces.go

```diff
@@ -0,0 +1,197 @@
+			if err != nil {
+				return nil, err
+			}
+			txs = append(txs, nestedCalls...)
+		}
+	}
+
+	return txs, nil
+}
+
+// transformToRevertedTx converts the result of call tracer into the reverted transaction information according to the KAS database scheme.
+func transformToRevertedTx(trace *vm.InternalTxTrace, block *types.Block, motherTx *types.Transaction) (*RevertedTx, error) {
+	return &RevertedTx{
+		TransactionHash: motherTx.Hash().Bytes(),
+		BlockNumber:     block.Number().Int64(),
+		RevertMessage:   trace.Reverted.Message,
+		ContractAddress: trace.Reverted.Contract.Bytes(),
+		Timestamp:       block.Time().Int64(),
+	}, nil
+}
+
+// transformToTraceResults converts the chain event to internal transactions as well as reverted transactions.
+func transformToTraceResults(event blockchain.ChainEvent) ([]*Tx, []*RevertedTx, error) {
+	var (
+		internalTxs []*Tx
+		revertedTxs []*RevertedTx
+	)
+	for txIdx, trace := range event.InternalTxTraces {
+		if !isInternalTxResult(trace) {
+			continue
+		}
+
+		tx := event.Block.Transactions()[txIdx]
+		receipt := event.Receipts[txIdx]
+
+		motherTx := getMotherTx(event.Block, txIdx, tx)
+		offset := int64(0)
+
+		// transforms the result into internal transaction which is associated with KLAY transfer recursively.
+		if receipt.Status == types.ReceiptStatusSuccessful {
+			internalTx, err := transformToInternalTx(trace, &offset, motherTx, true)
+			if err != nil {
+				logger.Error("Failed to transform tracing result into internal tx", "err", err)
+				return nil, nil, err
+			}
+			internalTxs = append(internalTxs, internalTx...)
+		}
+
+		// transforms the result into an evm reverted transaction.
+		if receipt.Status == types.ReceiptStatusErrExecutionReverted {
+			revertedTx, err := transformToRevertedTx(trace, event.Block, tx)
+			if err != nil {
+				logger.Error("Failed to transform tracing result into reverted tx", "err", err)
+				return nil, nil, err
+			}
+			revertedTxs = append(revertedTxs, revertedTx)
+		}
+	}
+	return internalTxs, revertedTxs, nil
+}
+
+// InsertTraceResults inserts internal and reverted transactions in the given chain event into KAS database.
+func (r *repository) InsertTraceResults(event blockchain.ChainEvent) error {
+	internalTxs, revertedTxs, err := transformToTraceResults(event)
+	if err != nil {
+		logger.Error("Failed to transform the given event to tracing results", "err", err, "blockNumber", event.Block.NumberU64())
+		return err
+	}
+
+	if err := r.insertTransactions(internalTxs); err != nil {
+		logger.Error("Failed to insert internal transactions", "err", err, "blockNumber", event.Block.NumberU64(), "numInternalTxs", len(internalTxs))
+		return err
+	}
+
+	if err := r.insertRevertedTransactions(revertedTxs); err != nil {
+		logger.Error("Failed to insert reverted transactions", "err", err, "blockNumber", event.Block.NumberU64(), "numRevertedTxs", len(revertedTxs))
+		return err
+	}
+	return nil
+}
```

Comment:
Please add metrics for insertion performances.

Reply:
Agreed. Can I do this as another task? It seems that adding many more metrics is necessary.

Reply:
Understood. Please make a JIRA issue to track this.